### ISSUE TITLE
Offline playback / Sync

### ISSUE THREAD

**Issue by DarkShadowxx** [jeffvli/feishin#47]
Hi, 
I often play my music in places with no or bad internet access.

It would be nice to be able to download playlists and artists.
Maybe even with format/bitrate selection (to save storage space) and a download overview page?

**Comment by 4zv4l** [comment]
I agree, an offline mode could be great !

**Comment by foinf** [comment]
I really want to use this to access my navidrome on a windows machine but the lack of offline caching is a big hurdle.

**Comment by emifloofy** [comment]
If this was to get added, I feel like it would be greatly appreciated if there was an option to transcode all the songs that are stored locally. My navidrome server has all the songs in flac quality, and I feel like my laptop that's still rocking nothing but a single 256GB drive would really appreciate it if I'd cut it some slack this way... :sweat_smile:

**Comment by Gecked-Deck** [comment]
Boost on this. I really want to use Feishin on my Laptop but I'm really often without stable internet,

**Comment by raphmim** [comment]
Hello, I'm testing out Feishin and it all looks great but without offline support it's unusable to me :(

**Comment by esiqveland** [comment]
Could we use the relatively new FileSystem API to start storing offline data for the tracks at least?

https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory

**Comment by esiqveland** [comment]
> Could we use the relatively new FileSystem API to start storing offline data for the tracks at least?
> 
> https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory

I did a minimal experiment with StorageManager API, and I think it can be done using StorageManager.

The most unclear parts are:
- Need a way to playback the data. I dont think we can run a proxy webserver to serve the files, as the file data would be locked away inside StorageManager of Chromium. But youtube somehow makes it work for them.
- Need a persistent way to keep track of which songs we have downloaded or are supposed to hold offline. IndexedDB ?
  - Track formats and transcoding of format if needed.
- The metadata will also need to be stored offline for full offline usage.
  - Plus album art

So to make it truly offline will be quite a bit of work.

**Comment by ak1932** [comment]
Is this something I can work on? If so, could someone direct me to the parts that would be required to change.

**Comment by nellePoint** [comment]
> Is this something I can work on? If so, could someone direct me to the parts that would be required to change.

its open source, you can fork it, look at the source, make changes and submit a pull request

**Comment by tkr-sh** [comment]
@nellePoint This is not that easy. This project is missing a CONTRIBUTING.md and we don't know if the owner is ok to have this feature.
I might look into it in the future since it seems important to me

**Comment by tkr-sh** [comment]
@CatOfPower Do you know all the time and investment it takes to make an maintain an open source project on years ?
Sure you can do it. Like you can also write it from scratch if you want too.
But suggesting that is completely unreasonable and not helping the original author of the comment _at all_.

**Comment by x86-unofficial** [comment]
Since this doesn't seem to be going anywhere, may I just suggest anyone who is using a jellyfin backend and need offline playback on mobile to try another client https://github.com/jmshrv/finamp.
I've been using it's offline playback for years on iOS and it's working perfectly.

**Comment by hansyulian** [comment]
Has anyone started doing this? if not i'm going to make this function. I also need this as i have been quite mobile a lot, i realized that once i am outside, my macbook lose all its music

**Comment by ak1932** [comment]
@dfrctn-sprn0va I don't think it is very nice to suggest another project (open source or not) on an open source project issue page where people are not paid to contribute, especially to a potential contributor.

**Comment by adri99-net** [comment]
I agree. I'm a mega-fan of Feishin, and offline playback would be wonderful for traveling.

**Comment by Statick** [comment]
+1 for this, being able to access some or all of my library offline is exactly what I need, storing a manual copy and using a different player is annoying

**Comment by raphmim** [comment]
@jeffvli what do you think about this feature request? Any plan to work on it?

**Comment by jeffvli** [comment]
> [@jeffvli](https://github.com/jeffvli) what do you think about this feature request? Any plan to work on it?

I still don't have any immediate plans for implementation.

I've been slowly reworking parts of the app into a cleaner architecture to support this, but until I can stabilize Feishin's data model I don't want to touch any offline implementation. 

Any work on this will be after Navidrome releases its stable API, after which I can pivot back here.

**Comment by theacodes** [comment]
@jeffvli is there a tracking issue for the Navidrome API? Which parts of it are needed for offline sync?

**Comment by jeffvli** [comment]
> is there a tracking issue for the Navidrome API? 

Not that I know of, but I will update this issue if/once I have any progress on it.

> Which parts of it are needed for offline sync?

 It's not that I *need* the new API, it's just that I don't want to work on anything related to persisting and caching data until I know that the data model is not going to immediately need to change.

**Comment by theacodes** [comment]
Word, thank you.

**Comment by Rainoutt** [comment]
Did Navidrome already updated its API?

**Comment by jeffvli** [comment]
> Did Navidrome already updated its API?

No, it's still TBD.

It was actively being worked on the last I heard, although the Navidrome developer is currently on break for a few weeks.

**Comment by Dr4goonD3mon** [comment]
I think Offline playback is needed, I live where I can't have internet in all moments and when there's an energy blackout my NAS is off so I couldn't play my music in my laptop, the phone can, but I want to have this on my laptop

**Comment by hi12167pies** [comment]
would love to see this
